From ff99091eb8a6b9e6edc567f6d2552183fbaacec3 Mon Sep 17 00:00:00 2001
From: Wenlin Kang <wenlin.kang@windriver.com>
Date: Mon, 6 Nov 2023 14:43:28 +0800
Subject: [PATCH] lib/util-get-random-bytes.c: fixed conversion error with
 mingw

With x86_64-w64-mingw32-gcc. get below error:
| ../git/lib/util-get-random-bytes.c: In function '_crypt_get_random_bytes':
| ../git/lib/util-get-random-bytes.c:140:42: error: conversion from 'size_t' {aka 'long long unsigned int'} to 'unsigned int' may change value [-Werror=conversion]
|   140 |           ssize_t nread = read (fd, buf, buflen);
|       |                                          ^~~~~~

In util-get-random-bytes.c, has get_random_bytes(void *buf, size_t buflen),
but in mingw-w64-mingw-w64/mingw-w64-headers/crt/io.h, read() has "unsigned int"
read(int _FileHandle,void *_DstBuf,unsigned int _MaxCharCount), and has:
 #ifdef _WIN64
   __MINGW_EXTENSION typedef unsigned __int64 size_t;
 #else
   typedef unsigned int size_t;
 #endif /* _WIN64 */

Upstream-Status: Pending

Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 lib/util-get-random-bytes.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/lib/util-get-random-bytes.c b/lib/util-get-random-bytes.c
index 79816db..68cd378 100644
--- a/lib/util-get-random-bytes.c
+++ b/lib/util-get-random-bytes.c
@@ -137,7 +137,11 @@ get_random_bytes(void *buf, size_t buflen)
         dev_urandom_doesnt_work = true;
       else
         {
+#ifdef _WIN64
+          ssize_t nread = read (fd, buf, (unsigned int)buflen);
+#else
           ssize_t nread = read (fd, buf, buflen);
+#endif
           if (nread < 0 || (size_t)nread < buflen)
             dev_urandom_doesnt_work = true;
 
-- 
2.25.1

